src/datasources.js

```diff
@@ -30,142 +30,116 @@
-                    };
-                newBaseFixture.name = baseFixture.name + " - " + (value.name || key);
-
-                if (!value.no_option) {
-                    ret = [newBaseFixture];
-                }
-                return ret.concat(generateFixtureDefinitions(value.structure, newBaseFixture));
-            });
-        }
-
-        return _.flatten(_.map(cachedDataSources.fixture, function(fixture) {
-            var baseFixture = {
-                src: fixture.sourceUri,
-                id: fixture.defaultId,
-                query: fixture.initialQuery,
-                name: fixture.name || fixture.defaultId
-            };
-
-            return [baseFixture].concat(generateFixtureDefinitions(fixture.structure, baseFixture));
-        }));
+        reset();
     }
 
-    function generateFixtureOptions() {
-        return _.map(getPossibleFixtures(), function(fixture) {
-            return {
-                value: JSON.stringify(_.omit(fixture, 'name')),
-                text: fixture.name
-            };
-        });
-    }
-
-    function generateFixtureColumns(fixture) {
-        function generateColumns(structure) {
-            return _.map(structure, function(value, key) {
-                return [key].concat(_.map(generateColumns(value.structure), function(value) {
-                    return key + '/' + value;
-                }));
-            });
-        }
-
-        if (fixture) {
-            return _.flatten(generateColumns(fixture.structure));
-        }
-        return [];
-    }
-
-    function autocompleteChoices(fixture_uri) {
-        return generateFixtureColumns(cachedDataSources.fixture[fixture_uri]);
+    function reset() {
+        dataCache = {};
+        dataCallbacks = {};
     }
 
+    /**
+     * Asynchronously load data sources of the given type
+     *
+     * @param type - The data source type (example: "fixture").
+     * @param callback - A function to be called when the data sources
+     *      have been loaded. This function should accept two arguments:
+     *      data - {
+     *               <sourceUri1>: <data source 1 object>,
+     *               <sourceUri2>: <data source 2 object>,
+     *               ...
+     *             }
+     *      type - the data source type.
+     */
     function getDataSources(type, callback) {
-        if (!_.isEmpty(cachedDataSources[type])) {
+        if (!_.isUndefined(dataCache[type])) {
+            callback(dataCache[type], type);
+            return;
+        }
+        if (!_.isUndefined(dataCallbacks[type])) {
+            dataCallbacks[type].push(callback);
             return;
         }
 
+        function finish(data) {
+            dataCache[type] = {};
+            if (data.length === 0) {
+                dataCache[type][""] = {
+                    sourceUri: "",
+                    defaultId: "",
+                    initialQuery: "",
+                    name: "Not Found",
```

Comment:
https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/templates/app_manager/form_designer.html#L25 ?

Could change that to lookup tables, but that would also label products as lookup tables right now